apps/service_providers/llm_service/adapters.py

```diff
@@ -119,17 +93,176 @@ def save_message_to_history(self, message: str, type_: ChatMessageType, experime
         if type_ == ChatMessageType.AI:
             self.ai_message = chat_message
             chat_message.add_version_tag(
-                version_number=self.experiment.version_number, is_a_version=self.experiment.is_a_version
+                version_number=self.experiment_version_number, is_a_version=self.experiment_is_a_version
             )
 
+        return chat_message
+
+
+class ChatAdapter(BaseAdapter):
+    def __init__(
+        self,
+        session: ExperimentSession,
+        provider_model_name: str,
+        llm_service,
+        temperature: int,
+        prompt_text: str,
+        max_token_limit: int,
+        experiment_version_number: int,
+        experiment_is_a_version: bool,
+        tools: list = None,
+        input_formatter: str | None = None,
+        source_material_id: str | None = None,
+        trace_service=None,
+    ):
+        self.session = session
+        self.team = session.team
+        self.provider_model_name = provider_model_name
+        self.llm_service = llm_service
+        self.temperature = temperature
+        self.prompt_text = prompt_text
+        self.max_token_limit = max_token_limit
+        self.experiment_version_number = experiment_version_number
+        self.experiment_is_a_version = experiment_is_a_version
+        self.tools = tools or []
+        self.input_formatter = input_formatter
+        self.source_material_id = source_material_id
+
+        self.trace_service = trace_service
+        self.template_context = PromptTemplateContext(session, source_material_id)
+
+    @staticmethod
+    def from_experiment(experiment: Experiment, session: ExperimentSession, trace_service=None) -> Self:
+        return ChatAdapter(
```

Comment:
This should be a classmethod and use `cls` as the constructor (https://typing.readthedocs.io/en/latest/spec/generics.html#use-in-classmethod-signatures):

```suggestion
    @classmethod
    def from_experiment(cls, experiment: Experiment, session: ExperimentSession, trace_service=None) -> Self:
        return cls(
```

Same for the other ones